<?xml version="1.0"?>
<!--
Copyright (c) 2013, Goncalo Cabrita, ISR University of Coimbra
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of Clearpath Robotics nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

<robot xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor"
	xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller"
	xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface"
	xmlns:xacro="http://ros.org/wiki/xacro" name="fsr_husky_arm">
<xacro:macro name="fsr_husky_arm" params="parent x y z rx ry rz"> 
    
  <link
    name="base_arm_link">
    <inertial>
      <origin
        xyz="0.062584 0.18878 0.37271"
        rpy="0 0 0" />
      <mass
        value="2.5803" />
      <inertia
        ixx="0.0087778"
        ixy="-0.00034853"
        ixz="0.0010654"
        iyy="0.014921"
        iyz="-0.0021847"
        izz="0.01488" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/base_arm_link.STL" />
      </geometry>
      <material name="Black" />
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/base_arm_link.STL" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="base_arm_link">
        <material>Gazebo/Black</material>
	<selfCollide>false</selfCollide>
        <turnGravityOff>true</turnGravityOff>
    </gazebo>

  <joint name="arm_joint" type="fixed">
			<parent link="${parent}"/>
			<child link="base_arm_link"/>
			<origin xyz="${x} ${y} ${z}" rpy="${rx} ${ry} ${rz}" /> 
  </joint>

  <link
    name="arm_axel">
    <inertial>
      <origin
        xyz="-0.0045492 0.35516 -0.025701"
        rpy="0 0 0" />
      <mass
        value="0.49315" />
      <inertia
        ixx="0.0060272"
        ixy="-0.00056478"
        ixz="8.6784E-05"
        iyy="0.0010748"
        iyz="-0.00058695"
        izz="0.0064909" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/arm_axel.STL" />
      </geometry>
      <material
        name="DarkGray">
        <color
          rgba="0.776470588235294 0.756862745098039 0.737254901960784 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/arm_axel.STL" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="arm_axel">
        <material>Gazebo/Grey</material>
        <selfCollide>false</selfCollide>
        <turnGravityOff>true</turnGravityOff>
    </gazebo>

  <joint
    name="arm_axel_joint"
    type="revolute">
    <origin
      xyz="0.12828 -0.4402 0.049088"
      rpy="1.5708 6.123E-17 1.5708" />
    <parent
      link="base_arm_link" />
    <child
      link="arm_axel" />
    <axis
      xyz="0 -1 0" />
    <limit
      lower="-1.5708"
      upper="3.06"
      effort="75"
      velocity="2" />
  </joint>

  <link
    name="upper_arm">
    <inertial>
      <origin
        xyz="9.76184988289996E-05 0.0241578589661459 0.355397722899461"
        rpy="0 0 0" />
      <mass
        value="0.326325490126923" />
      <inertia
        ixx="0.0322020258970473"
        ixy="-4.4638355127408E-08"
        ixz="1.53741858406116E-05"
        iyy="0.032199356997134"
        iyz="-0.000144641063316242"
        izz="9.23314207900664E-05" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/upper_arm.STL" />
      </geometry>
      <material
        name="DarkGray">
        <color
          rgba="0.776470588235294 0.756862745098039 0.737254901960784 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/upper_arm.STL" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="upper_arm">
        <material>Gazebo/Grey</material>
        <selfCollide>false</selfCollide>
        <turnGravityOff>true</turnGravityOff>
    </gazebo>

  <joint
    name="upper_arm_joint"
    type="revolute">
    <origin
      xyz="0 0.44 0.011736"
      rpy="1.5708 0 -1.5708" />
    <parent
      link="arm_axel" />
    <child
      link="upper_arm" />
    <axis
      xyz="0 -1 0" />
    <limit
      lower="-0.5236"
      upper="0.16"
      effort="35"
      velocity="2" />
  </joint>

  <link
    name="metal_detector_arm">
    <inertial>
      <origin
        xyz="-0.0131202075717661 -0.0723031614133108 -0.0463889610764896"
        rpy="0 0 0" />
      <mass
        value="0.20115578586952" />
      <inertia
        ixx="0.00493443507445138"
        ixy="-0.000620294299504152"
        ixz="-0.000306843036961987"
        iyy="0.0014212633147968"
        iyz="0.00125817384513895"
        izz="0.00527185766074029" />
    </inertial>
    <visual>
      <origin
        xyz="0.0255 0.0026 -0.01050"
        rpy="-1.865 -0.1 -1" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/metal_detector_arm.STL" />
      </geometry>
      <material
        name="DarkGray">
        <color
          rgba="0.494117647058824 0.494117647058824 0.494117647058824 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0.0255 0.0026 -0.01050"
        rpy="-1.865 -0.1 -1" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/metal_detector_arm.STL" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="metal_detector_arm">
        <material>Gazebo/Grey</material>
        <selfCollide>false</selfCollide>
        <turnGravityOff>true</turnGravityOff>
    </gazebo>

  <joint
    name="metal_detector_arm_joint"
    type="revolute">
    <mimic
      joint="upper_arm_joint"
      multiplier="-1" />
    <origin
      xyz="0.0 0.0344999999999998 0.8483340615005693"
      rpy="-1.5707963267949 1.140744854051 -1.26098072342103" />
    <parent
      link="upper_arm" />
    <child
      link="metal_detector_arm" />
    <axis
      xyz="0.397068061333986 -0.865669276483222 -0.30488302416088" />
	<limit
      lower="-0.5236"
      upper="0.16"
      effort="35"
      velocity="2" />
  </joint>

  <link
    name="metal_detector_antenna">
    <inertial>
      <origin
        xyz="0.124781096368378 -0.279414752184092 -0.122957072146209"
        rpy="0 0 0" />
      <mass
        value="0.80992798706323" />
      <inertia
        ixx="0.223638341370907"
        ixy="0.0140547266307369"
        ixz="0.0258666395393033"
        iyy="0.250461845552196"
        iyz="-0.0571594766057374"
        izz="0.0805524351506661" />
    </inertial>
    <visual>
      <origin
        xyz="0.026 0 -0.29"
        rpy="-1.9 -0.45 -1.45" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/metal_detector_antenna.STL" />
      </geometry>
      <material
        name="ArmyGreen">
        <color
          rgba="0.3 0.332 0.128 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0.026 0 -0.29"
        rpy="-1.9 -0.45 -1.45" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/metal_detector_antenna.STL" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="metal_detector_antenna">
        <material>Gazebo/Green</material>
        <selfCollide>false</selfCollide>
        <turnGravityOff>true</turnGravityOff>
    </gazebo>

  <joint
    name="metal_detector_antenna_joint"
    type="fixed">
    <origin
      xyz="0.0350 -0.075 0.27"
      rpy="0.350 0 0.42" /> -->
    <parent
      link="metal_detector_arm" />
    <child
      link="metal_detector_antenna" />
    <axis
      xyz="0 0 0" />
  </joint>
  
  <link name="left_coil">
      <visual>
          <origin xyz="0 0 0" rpy="0 0 0" />
          <geometry>
              <box size="0.001 0.001 0.001" />
          </geometry>
      </visual>
  </link>
  
  <joint name="left_coil_joint" type="fixed">
      <origin xyz="0.18 -0.02 0.04" rpy="0 0 0" />
      <parent link="metal_detector_antenna"/>
      <child link="left_coil" />
  </joint>
  
  <link name="middle_coil">
      <visual>
          <origin xyz="0 0 0" rpy="0 0 0" />
          <geometry>
              <box size="0.001 0.001 0.001" />
          </geometry>
      </visual>
  </link>
  
  <joint name="middle_coil_joint" type="fixed">
      <origin xyz="0 -0.02 0.04" rpy="0 0 0" />
      <parent link="metal_detector_antenna"/>
      <child link="middle_coil" />
  </joint>
  
  <link name="right_coil">
      <visual>
          <origin xyz="0 0 0" rpy="0 0 0" />
          <geometry>
              <box size="0.001 0.001 0.001" />
          </geometry>
      </visual>
  </link>
  
  <joint name="right_coil_joint" type="fixed">
      <origin xyz="-0.18 -0.02 0.04" rpy="0 0 0" />
      <parent link="metal_detector_antenna"/>
      <child link="right_coil" />
  </joint>


  <link
    name="lower_arm">
    <inertial>
      <origin
        xyz="-0.234602780446024 -0.225204746125893 0.0031580399460546"
        rpy="0 0 0" />
      <mass
        value="0.326325490126923" />
      <inertia
        ixx="0.0167023392786261"
        ixy="0.0172620654204666"
        ixz="0.00171207285568604"
        iyy="0.0255068597107398"
        iyz="0.000242235271108958"
        izz="0.0382877611003989" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/lower_arm.STL" />
      </geometry>
      <material
        name="DarkGray">
        <color
          rgba="0.776470588235294 0.756862745098039 0.737254901960784 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://hratc2015_framework/models/arm/lower_arm.STL" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="lower_arm">
        <material>Gazebo/Grey</material>
        <selfCollide>false</selfCollide>
        <turnGravityOff>true</turnGravityOff>
    </gazebo>

  <joint
    name="lower_arm_joint"
    type="revolute">
    <mimic
      joint="upper_arm_joint" />
    <origin
      xyz="0 0.375 0"
      rpy="3.1416 0 0.70689" />
    <parent
      link="arm_axel" />
    <child
      link="lower_arm" />
    <axis
      xyz="0 0 1" />
    <limit
      lower="-0.5236"
      upper="0.16"
      effort="35"
      velocity="2" />
	</joint>
  
  <!-- Transmissions -->
  <transmission name="lift">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="upper_arm_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="lift_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  
  <transmission name="mimic_lift1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="lower_arm_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="mimic_lift_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  
  <transmission name="mirror_lift2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="metal_detector_arm_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="mirror_lift_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="sweep">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="arm_axel_joint">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="sweep_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  
  <!-- Gazebo mimic plugins -->
  <gazebo>
	<plugin name="mimic_plugin1" filename="libgazebo_mimic_plugin.so">
            <joint>upper_arm_joint</joint>
            <mimicJoint>lower_arm_joint</mimicJoint>
            <multiplier>1.0</multiplier>
        </plugin>
  </gazebo>
  
  <gazebo>
	<plugin name="mimic_plugin2" filename="libgazebo_mimic_plugin.so">
            <joint>upper_arm_joint</joint>
            <mimicJoint>metal_detector_arm_joint</mimicJoint>
            <multiplier>-1.0</multiplier>
        </plugin>
  </gazebo>
  
</xacro:macro>
</robot>
